### ISSUE TITLE
​DetailsList: Resize columns behavior changed after strict mode fixes

### ISSUE THREAD

**Issue by clavx** [microsoft/fluentui#15971]
### Environment Information

- **Package version(s)**: 7.149.4
- **Browser and OS versions**: Windows 10 + Edge

### Describe the issue:

The behavior of the header column of DetailsList changed when it was updated from v7.140.0 to v7.141.0.

7.141.0 or later
- Double-clicking on a column to AutoResize only allocates space in the display area and does not produce a horizontal scroll bar.
- Cannot view the contents that don't fit in the column.

7.140.0 or earlier
- Double-clicking on a column to AutoResize widens it out of the display area and produces a horizontal scroll bar.
- Can view the contents that doesn't fit in the column.

I want to keep the behavior as it was before 7.140.0, how should I set up the DetailsList?

#### Actual behavior:

<img src="https://user-images.githubusercontent.com/58804032/99528349-81621000-29e1-11eb-80ce-06b253b0eb22.gif" width="600">

#### Expected behavior:

<img src="https://user-images.githubusercontent.com/58804032/99528361-8757f100-29e1-11eb-8076-2f53026c18ba.gif" width="600">

**Comment by paulgildea** [comment]
@clavx Thanks for filing this issue and providing a lot of detail on the issue! This is super helpful for us. We'll take a look.

**Comment by ecraig12345** [comment]
This appears to be caused by @ThomasMichon's PR #14980 which added strict mode support to DetailsList. #16332 from @dps88 is another resizing-related issue introduced in the same version, with slightly different steps and desired behavior but I suspect the same root cause.

I can't easily verify @clavx's scenario directly since there's no codepen (please provide a codepen link in the future, not just screenshots), but I was able to verify that the code from #16332 [works in 7.140.0](https://codepen.io/ecraig12345/pen/xxEovBB), but not in the [PR deploy site for #14980](http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/14980/merge/fabric-website-resources/dist/index.html#/examples/detailslist/basic). (Click "Show code" on the example on that page, then paste in the codepen contents with a couple modifications noted in comments.)

I'm going to close #16332 as a duplicate of this one, but copying the description below since it provides some helpful additional info.

----

### Environment Information

- **Package version(s)**: "@fluentui/react": "^7.155.3"

### Please provide a reproduction of the bug in a codepen:  
https://codepen.io/duschroe/pen/poEawVV

#### Has this been discussed before?:

Roughly, Yes. [this issue] #15971, #15786.  I think those are missing some helpful information though.

#### Actual behavior:

When a details list has `ConstrainMode.unconstrained` and `DetailsListLayoutMode.justified`, the way columns that overflow are handled is a little inconsistent.

- If no columns are provided, overflow columns drop off the table.

- If columns are provided, overflow columns remain
   - Adding `isCollapsible: true` can be used to match the default columns
   - However, without `isCollapsible`, overflow columns cannot be resized once they are in overflow.
   - This isn't good for tables with many columns and/or multiline columns. 

- Only the first column is resizable if the combined column width is larger than the table viewport

#### Expected behavior:

Using fixed columns can be used for most cases where there are too many columns but doesn't work great for few columns or really large browser viewports. It creates wasted space  
![image](https://user-images.githubusercontent.com/50461997/103431257-41ff1580-4b82-11eb-8e4d-765cf9b464e8.png)

With `ConstrainMode.unconstrained` and `DetailsListLayoutMode.justified`, ideally a Details list should:

- Do it's best to scale the columns appropriately if columns fit in viewport

- If they don't fit, use minWidth and *observe user resizing for all columns.  Even if the column is in overflow*.
  - The DetailsList was working roughly like that when we started using FluentUI (roughly 7.110) and is the case if you change the version in the codepen I provided.

- Maintain the resized column width across all columns
  - Earlier versions reset overflow columns to minWidth if they are not the column being currently adjusted

**Comment by clavx** [comment]
Sorry,

Actual behavior: v7.141.0
https://codepen.io/clavx/pen/rNWOVXY

Expected behavior: v7.140.0
https://codepen.io/clavx/pen/poNjJYR

**Comment by ThomasMichon** [comment]
Unfortunately, my take on this issue is that the behavior you see from *before* the `StrictMode` support work *was itself a bug*. All the `StrictMode` work did was impose a little more consistency on the justification behavior, so that it also applies as columns are resizing and not just when the viewport width is changed.

One bug remaining is that the left-most column can still be resized larger, even if all columns to the right are at their minimum widths and the total width is larger than the viewport width. This is due to a faulty index check which carves out an exception to the justification behavior for the first column.

If we want to have a behavior where all columns are resizable if the *minimum total width* still exceeds the viewport width, we must do so explicitly. That behavior needs to be specifically called out in the justification logic, where it essentially 'shuts off' in that case. That's not hard to do and I can produce a PR witha  fix shortly.

**Comment by sysofwan** [comment]
@ThomasMichon Thanks for the PR. Looks like there is some test failure. Can you complete it?

**Comment by sysofwan** [comment]
@ThomasMichon Thanks for the fix! The issue is also in v7 branch. Can you help cherry-pick the fix?

**Comment by makislev** [comment]
Is there any update on the cherry pick to v7 branch? we are waiting on this fix as blocker to open a new feature to customers
